src/pages/topology/Regions.js

```diff
@@ -0,0 +1,139 @@
+* limitations under the License.
+**/
+import React, { Component } from 'react';
+import { HashLink } from 'react-router-hash-link';
+import { translate } from '../../localization/localize.js';
+import '../../styles/deployer.less';
+import { getInternalModel } from './TopologyUtils.js';
+
+/*
+ * This class is a JavaScript implementation of the script
+ * ardana_configurationprocessor/plugins/builders/HTMLDiagram/ControlPlanes.py
+ * in the config processor
+ */
+class Regions extends Component {
+  constructor(props) {
+    super(props);
+
+    this.state = {
+      model: undefined
+    };
+
+  }
+
+  componentDidMount() {
+
+    getInternalModel()
+      .then((yml) => {
+
+        // Force a re-render if the page is still shown (user may navigate away while waiting)
+        if (this.refs.regions)
+          this.setState({
+            model: yml});
+      })
+      .catch((error) => {
+        console.log(error); // eslint-disable-line no-console
+      });
+  }
+
+  render_regions = () => {
+
+    const cps = this.state.model['internal']['control-planes'];
+
+    // Build a map by region of components advertised to keystone.  Note
+    // that the component name will be the specific service, e.g. cinder-api
+    let advertised = {};
+    for (const cp of Object.values(cps)) {
+      const cp_advertised = cp['advertised'] || {};
+      for (const adv of Object.values(cp_advertised['keystone-api'] || [])) {
+        for (const region of adv['regions']) {
+          advertised[region] = advertised[region] || [];   // Seed with an empty array
+          advertised[region].push(adv['component_name']);
+        }
+      }
+    }
+    const regions = this.state.model['internal']['region-topology']['regions'];
+
+    // Build the header row of the table
+    let header_row = Object.keys(regions).sort().map(name => <th key={name}>{name}</th>);
+    header_row.unshift(<th key="cp">{translate('control_planes')}</th>);
+
+    // Build the data rows
+    const table_rows = Object.keys(cps).sort().map(cp_name => {
+
+      const columns = Object.keys(regions).sort().map(name => {
+        if (regions[name]['control_planes'][cp_name]) {
+
+          // region_svcs is an object with keys corresponding to service names (e.g. 'cinder'),
+          // and whose values are arrays of component names (e.g. 'cinder-api', 'cinder-schduler', etc.)
+          const region_svcs = regions[name]['control_planes'][cp_name]['services'];
+
+          const service_list = Object.keys(region_svcs).map(svc => {
+
+            // If any component name for this service is advertised to keystone, then
+            // highlight it in the displayed list
+            let found = false;
+            for(const component of region_svcs[svc]) {
+              if (advertised[name].includes(component)) {
+                found = true;
+                break;
+              }
+            }
+            if (found) {
+              return <div key={svc}><b>{svc}</b></div>;
```

Comment:
the found one in bold with other one in slightly lighter found looks a little odd at the first sight. Might need UX input for that to make it obvious to user that one is advertised through keystone  and one is not.

Reply:
I agree that this way of using bold to represent items advertised by keystone is not great -- I was simply reproducing here what was done in the original pages produced by the config processor. It, as well as all of these topology pages, needs some UX help.

Reply:
agree,  thanks.